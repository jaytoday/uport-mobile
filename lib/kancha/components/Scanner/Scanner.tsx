import * as React from 'react'
import { ViewStyle, SafeAreaView } from 'react-native'
import { Container, Button, Text, Icon } from '@kancha'
import { RNCamera } from 'react-native-camera'

interface ScannerProps {
  /**
   * Callback when barcode has been read
   */
  onBarcodeRead: (event: any) => void
  /**
   * Close the scanner
   */
  closeScanner: () => void

  /**
   * Camera has permission
   */
  hasPermission: boolean | null
}

const NoPermission: React.FC<{ closeScanner: () => void }> = ({ closeScanner }) => {
  return (
    <Container flex={1} backgroundColor={'black'} alignItems={'center'} justifyContent={'center'} padding={20}>
      <Container paddingBottom>
        <Text type={Text.Types.H2} textColor={'white'} textAlign={'center'}>
          Connect with uPort
        </Text>
      </Container>
      <Text type={Text.Types.SubTitle}>
        QR scanner requires camera access. Enable camera access in settings so you can scan QR codes and start
        connecting to apps.
      </Text>
      <Container marginTop={40}>
        <Button
          type={Button.Types.Secondary}
          onPress={() => closeScanner()}
          block={Button.Block.Clear}
          buttonText={'Close'}
        />
      </Container>
    </Container>
  )
}

const QRCodeScanner: React.FC<ScannerProps> = ({ onBarcodeRead, closeScanner, hasPermission }) => {
  const scannerViewStyles: ViewStyle = {
    flex: 1,
    justifyContent: 'flex-end',
    alignItems: 'center',
  }

  return hasPermission ? (
    <RNCamera captureAudio={false} onBarCodeRead={onBarcodeRead} style={{ flex: 1 }}>
      <SafeAreaView style={scannerViewStyles}>
        <Container paddingBottom={35}>
          <Button
            type={Button.Types.Secondary}
            onPress={() => closeScanner()}
            block={Button.Block.Clear}
            buttonText={'Close'}
          />
        </Container>
      </SafeAreaView>
    </RNCamera>
  ) : (
    <NoPermission closeScanner={closeScanner} />
  )
}

export default QRCodeScanner
