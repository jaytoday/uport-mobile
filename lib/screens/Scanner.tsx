import * as React from 'react'
import { Vibration, AppState } from 'react-native'
import Permissions from 'react-native-permissions'
import { Navigation } from 'react-native-navigation'
import { Screen, Container, Scanner, Device, Theme } from '@kancha'

// Redux Connect
import { connect } from 'react-redux'
import { handleURL } from 'uPortMobile/lib/actions/requestActions'

/**
 * Timeout value for scanner to stop trying to scan QR codes
 */
const SCANNER_TIMEOUT = 10000

interface ScannerScreenProps {
  componentId: string
  handleQRCodeURL: (event: any) => void
}

interface ScannerScreenState {
  isEnabled: boolean
  appState: string
  hasPermission: null | boolean
}

export class ScannerScreen extends React.Component<ScannerScreenProps, ScannerScreenState> {
  timeout: any

  constructor(props: ScannerScreenProps) {
    super(props)

    /**
     * manage all camera state in container
     */
    this.state = {
      isEnabled: false,
      appState: AppState.currentState,
      hasPermission: null,
    }

    Navigation.events().bindComponent(this)

    this.toggleScannerMode = this.toggleScannerMode.bind(this)
    this.onBarCodeRead = this.onBarCodeRead.bind(this)
    this.closeScanner = this.closeScanner.bind(this)
  }

  componentDidMount() {
    AppState.addEventListener('change', this._handleAppStateChange)
  }

  async componentDidAppear() {
    this.toggleScannerMode(true)

    let status = await Permissions.check('camera')

    if (status === 'undetermined') {
      status = await Permissions.request('camera')
    }

    this.setState({ ...this.state, hasPermission: status === 'authorized' })
  }

  componentDidDisappear() {
    this.toggleScannerMode(false)
  }

  componentWillUnmount() {
    AppState.removeEventListener('change', this._handleAppStateChange)
  }

  _handleAppStateChange = (nextAppState: string) => {
    this.setState({ ...this.state, appState: nextAppState })
  }

  toggleScannerMode(enabled: boolean) {
    this.setState({
      ...this.state,
      isEnabled: enabled,
    })
  }

  onBarCodeRead(event: any) {
    Vibration.vibrate(400, false)
    this.props.handleQRCodeURL(event)
    this.toggleScannerMode(false)
    this.closeScanner()
  }

  toggleIOSDrawer() {
    Navigation.mergeOptions('Scanner', {
      sideMenu: {
        right: {
          visible: false,
        },
      },
    })
  }

  popAndroidScannerView() {
    Navigation.dismissAllModals()
  }

  closeScanner() {
    if (Device.isIOS) {
      this.toggleIOSDrawer()
    } else {
      setTimeout(() => {
        this.popAndroidScannerView()
      }, 250)
    }
  }

  render() {
    return (
      <Screen config={Screen.Config.NoScroll} type={Screen.Types.Primary}>
        <Container flex={1} backgroundColor={'#000000'}>
          {this.state.appState === 'active' && this.state.isEnabled && (
            <Scanner
              hasPermission={this.state.hasPermission}
              onBarcodeRead={this.onBarCodeRead}
              closeScanner={this.closeScanner}
            />
          )}
        </Container>
      </Screen>
    )
  }
}

const mapStateToProps = (state: any, ownProps: any) => ownProps

const mapDispatchToProps = (dispatch: any) => {
  return {
    handleQRCodeURL: (event: any) => {
      if (event.data) {
        dispatch(handleURL(event.data, { postback: true }))
      }
    },
  }
}

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(ScannerScreen)
