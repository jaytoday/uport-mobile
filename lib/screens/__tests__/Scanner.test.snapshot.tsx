import 'react-native'
import React from 'react'
import { ScannerScreen } from '../Scanner'
import { render } from 'react-native-testing-library'

jest.mock('react-native-navigation', () => {
  return {
    Navigation: {
      events: jest.fn().mockReturnValue({ bindComponent: jest.fn }),
      mergeOptions: jest.fn(),
    },
  }
})

it('renders a Scanner Screen', () => {
  const tree = render(<ScannerScreen componentId={'TEST'} handleQRCodeURL={jest.fn} />).toJSON()
  expect(tree).toMatchSnapshot()
})
