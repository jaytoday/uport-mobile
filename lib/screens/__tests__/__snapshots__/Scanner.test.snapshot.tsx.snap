// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a Scanner Screen 1`] = `
<View
  pointerEvents="auto"
  style={
    Object {
      "alignItems": undefined,
      "backgroundColor": "#FFFFFF",
      "borderRadius": undefined,
      "flex": 1,
      "flexDirection": undefined,
      "height": undefined,
      "justifyContent": undefined,
      "marginBottom": undefined,
      "marginLeft": undefined,
      "marginRight": undefined,
      "marginTop": undefined,
      "padding": undefined,
      "paddingBottom": undefined,
      "paddingHorizontal": undefined,
      "paddingLeft": undefined,
      "paddingRight": undefined,
      "paddingTop": undefined,
      "width": undefined,
    }
  }
>
  <View
    pointerEvents="auto"
    style={
      Object {
        "alignItems": undefined,
        "backgroundColor": "#000000",
        "borderRadius": undefined,
        "flex": 1,
        "flexDirection": undefined,
        "height": undefined,
        "justifyContent": undefined,
        "marginBottom": undefined,
        "marginLeft": undefined,
        "marginRight": undefined,
        "marginTop": undefined,
        "padding": undefined,
        "paddingBottom": undefined,
        "paddingHorizontal": undefined,
        "paddingLeft": undefined,
        "paddingRight": undefined,
        "paddingTop": undefined,
        "width": undefined,
      }
    }
  />
</View>
`;
